src/plots/command.js

```diff
@@ -0,0 +1,369 @@
+                        if(refBinding.traces[j] !== binding.traces[j]) {
+                            return false;
+                        }
+                    }
+                } else {
+                    return false;
+                }
+            } else {
+                if(binding.prop !== refBinding.prop) {
+                    return false;
+                }
+            }
+        }
+
+        binding = bindings[0];
+        var value = binding.value;
+        if(Array.isArray(value)) {
+            value = value[0];
+        }
+        if(bindingsByValue) {
+            bindingsByValue[value] = i;
+        }
+    }
+
+    return refBinding;
+};
+
+exports.createCommandObserver = function(gd, commandList, onchange) {
+    var cache = {};
+    var lookupTable = {};
+    var check, remove;
+    var enabled = true;
+
+    // Determine whether there's anything to do for this binding:
+    var binding;
+    if((binding = exports.hasSimpleAPICommandBindings(gd, commandList, lookupTable))) {
+        bindingValueHasChanged(gd, binding, cache);
+
+        check = function check() {
+            if(!enabled) return;
+
+            var container, value, obj;
+            var changed = false;
+
+            if(binding.type === 'data') {
+                // If it's data, we need to get a trace. Based on the limited scope
+                // of what we cover, we can just take the first trace from the list,
+                // or otherwise just the first trace:
+                container = gd._fullData[binding.traces !== null ? binding.traces[0] : 0];
+            } else if(binding.type === 'layout') {
+                container = gd._fullLayout;
+            } else {
+                return false;
+            }
+
+            value = Lib.nestedProperty(container, binding.prop).get();
+
+            obj = cache[binding.type] = cache[binding.type] || {};
+
+            if(obj.hasOwnProperty(binding.prop)) {
+                if(obj[binding.prop] !== value) {
+                    changed = true;
+                }
+            }
+
+            obj[binding.prop] = value;
+
+            if(changed && onchange) {
+                // Disable checks for the duration of this command in order to avoid
+                // infinite loops:
+                if(lookupTable[value] !== undefined) {
+                    disable();
+                    Promise.resolve(onchange({
+                        value: value,
+                        type: binding.type,
+                        prop: binding.prop,
+                        traces: binding.traces,
+                        index: lookupTable[value]
+                    })).then(enable, enable);
+                }
+            }
+
+            return changed;
+        };
+
+        var checkEvents = [
+            'plotly_relayout',
+            'plotly_redraw',
+            'plotly_restyle',
+            'plotly_update',
+            'plotly_animatingframe',
+            'plotly_afterplot'
+        ];
```

Comment:
@etpinard I refactored this to just hook into a list of existing events. This seems much simpler and more straightforward.